### DISCUSSION TITLE
HeaderFieldValueList in empty in outboundDecorator but other way is ok!

### DISCUSSION THREAD

**Discussion by ChuLiqiang** [resiprocate/resiprocate#356] [Q&A] [answered]
![image](https://github.com/resiprocate/resiprocate/assets/2747722/c6b3b9b2-e1a4-4066-a558-04974f4e8bca)

I can get callid by use this code：

`ControlMessage.strCallID = request.header(h_CallId).value().c_str();`

but when I use this code i find it is empty value？can you tell me how to fix that?
```
    bool bGetHeaderValue = false;
    bGetHeaderValue = getHeaderValue(request,h_CallId,ControlMessage.strCallID);
    if(bGetHeaderValue == false)
    {
        return false;
    }
    bool SipCtrlImp::getHeaderValue(const SipMessage& request,HeaderBase& headerBase,string& headerValue)
{
    if(request.exists(headerBase) == false)
    {
        return false;
    }
    Headers::Type headerType = headerBase.getTypeNum();
    if(headerType != Headers::UNKNOWN)
    {
        Data headerData;
        const HeaderFieldValueList* hfv = request.getRawHeader(headerType);
        for(HeaderFieldValueList::const_iterator it = hfv->begin(); it != hfv->end(); it++)
        {
            it->toShareData(headerData);
            headerValue = headerData.c_str();
            return true;
        }
    }
    return false;
}
```
can not run in this loop
`for(HeaderFieldValueList::const_iterator it = hfv->begin(); it != hfv->end(); it++)`

**Comment by sgodin** [comment, marked as answer]
The answer to this is quite complicated.  

From HeadedrFieldValueList.hxx:

   A HeaderFieldValueList contains a ParserContainer (after parsing occurs) or may only contain
   a vector of unparsed headers (if parsing hasn't occurred).  Depending on how the SipMessage was
   constructured, it's possible a HeaderFieldValueList doesn't contain any information in the 
   mHeaders vector, and all the data is stored solely in the ParserContainer.  If a ParserContainer
   is set, then it must be used for all future encoding.  Looking for null (or not) from 
   getParserContainer can be used to see which storage mechanism is being used.

   Most methods on this class deal with manipulation of the header vector (ie: empty, size, clear,
   push_back, pop_back, iterator, etc.).

   The encode method will either use the ParserContainer to encode (if set), or will encode from the
   header vector otherwise.

   In order to query all the raw header values regardless of their storage format use the 
   getNumHeaderValues and getHeaderValueByIndex methods as follows:

   for(size_t i = 0; i < hfvl->getNumHeaderValues(); i++)
   {
      Data headerValue;
      hfvl->getHeaderValueByIndex(i, headerValue);
      ....
   }

An example use can be see here:
https://github.com/resiprocate/resiprocate/blob/master/resip/recon/RemoteParticipant.cxx#L230

Scott
